File: ExtractBitlockerKeys.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File name          : ExtractBitlockerKeys.py
# Author             : Podalirius (@podalirius_)
# Date created       : 19 Sep 2023


import argparse
from sectools.windows.ldap import raw_ldap_query, init_ldap_session
from sectools.windows.crypto import nt_hash, parse_lm_nt_hashes
import os
import sys
import sqlite3
import json
import xlsxwriter
import re


VERSION = "1.1"


def export_json(options, results):
    print("[>] Exporting results to %s ... " % options.export_json, end="")
    sys.stdout.flush()
    basepath = os.path.dirname(options.export_json)
    filename = os.path.basename(options.export_json)
    if basepath not in [".", ""]:
        if not os.path.exists(basepath):
            os.makedirs(basepath)
        path_to_file = basepath + os.path.sep + filename
    else:
        path_to_file = filename
    f = open(path_to_file, "w")
    f.write(json.dumps(results, indent=4) + "\n")
    f.close()
    print("done.")


def export_xlsx(options, results):
    print("[>] Exporting results to %s ... " % options.export_xlsx, end="")
    sys.stdout.flush()
    basepath = os.path.dirname(options.export_xlsx)
    filename = os.path.basename(options.export_xlsx)
    if basepath not in [".", ""]:
        if not os.path.exists(basepath):
            os.makedirs(basepath)
        path_to_file = basepath + os.path.sep + filename
    else:
        path_to_file = filename
    workbook = xlsxwriter.Workbook(path_to_file)
    worksheet = workbook.add_worksheet()

    header_format = workbook.add_format({'bold': 1})
    header_fields = ["Computer FQDN", "Domain", "Recovery Key", "Volume GUID", "Created At", "Organizational Units"]
    for k in range(len(header_fields)):
        worksheet.set_column(k, k + 1, len(header_fields[k]) + 3)
    worksheet.set_row(0, 20, header_format)
    worksheet.write_row(0, 0, header_fields)

    row_id = 1
    for computerfqdn in results.keys():
        data = [
            computerfqdn,
            results[computerfqdn]["domain"],
            results[computerfqdn]["recoveryKey"],
            results[computerfqdn]["volumeGuid"],
            results[computerfqdn]["createdAt"],
            results[computerfqdn]["organizationalUnits"],
        ]
        worksheet.write_row(row_id, 0, data)
        row_id += 1
    worksheet.autofilter(0, 0, row_id, len(header_fields) - 1)
    workbook.close()
    print("done.")


def export_sqlite(options, results):
    print("[>] Exporting results to %s ... " % options.export_sqlite, end="")
    sys.stdout.flush()
    basepath = os.path.dirname(options.export_sqlite)
    filename = os.path.basename(options.export_sqlite)
    if basepath not in [".", ""]:
        if not os.path.exists(basepath):
            os.makedirs(basepath)
        path_to_file = basepath + os.path.sep + filename
    else:
        path_to_file = filename

    conn = sqlite3.connect(path_to_file)
    cursor = conn.cursor()
    cursor.execute("CREATE TABLE IF NOT EXISTS bitlocker_keys(fqdn VARCHAR(255), domain VARCHAR(255), recoveryKey VARCHAR(255), volumeGuid VARCHAR(255), createdAt VARCHAR(255), organizationalUnits VARCHAR(1024));")
    for computerfqdn in results.keys():
        cursor.execute("INSERT INTO shares VALUES (?, ?, ?, ?, ?, ?)", (
                computerfqdn,
                results[computerfqdn][0]["domain"],
                results[computerfqdn][0]["recoveryKey"],
                results[computerfqdn][0]["volumeGuid"],
                results[computerfqdn][0]["createdAt"],
                results[computerfqdn][0]["organizationalUnits"],
            )
        )
    conn.commit()
    conn.close()
    print("done.")


def get_domain_from_distinguished_name(distinguishedName):
    domain = None
    if "dc=" in distinguishedName.lower():
        distinguishedName = distinguishedName.lower().split(',')[::-1]

        while distinguishedName[0].startswith("dc="):
            if domain is None:
                domain = distinguishedName[0].split('=',1)[1]
            else:
                domain = distinguishedName[0].split('=', 1)[1] + "." + domain
            distinguishedName = distinguishedName[1:]

    return domain


def get_ou_path_from_distinguished_name(distinguishedName):
    ou_path = None
    if "ou=" in distinguishedName.lower():
        distinguishedName = distinguishedName.lower().split(',')[::-1]

        # Skip domain
        while distinguishedName[0].startswith("dc="):
            distinguishedName = distinguishedName[1:]

        while distinguishedName[0].startswith("ou="):
            if ou_path is None:
                ou_path = distinguishedName[0].split('=',1)[1]
            else:
                ou_path = ou_path + " --> " + distinguishedName[0].split('=',1)[1]
            distinguishedName = distinguishedName[1:]

        return ou_path
    else:
        return ou_path


def parse_fve(distinguishedName, bitlocker_keys):
    entry = {
        "distinguishedName": distinguishedName,
        "domain": get_domain_from_distinguished_name(distinguishedName),
        "organizationalUnits": get_ou_path_from_distinguished_name(distinguishedName),
        "createdAt": None,
        "volumeGuid": None
    }
    # Parse CN of key
    matched = re.match(r"^(CN=)([0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9]T[0-9][0-9]:[0-9][0-9]:[0-9][0-9]-[0-9][0-9]:[0-9][0-9])({[0-9A-F\-]+}),", distinguishedName, re.IGNORECASE)
    if matched is not None:
        _, created_at, guid = matched.groups()
        entry["createdAt"] = created_at
        entry["volumeGuid"] = guid.strip('{}').lower()
    # Parse computer name
    entry["computerName"] = None
    if ',' in distinguishedName:
        if distinguishedName.split(',')[1].upper().startswith("CN="):
            entry["computerName"] = distinguishedName.split(',')[1].split('=',1)[1]
    # Add recovery key
    entry["recoveryKey"] = bitlocker_keys["msFVE-RecoveryPassword"]

    return entry


def parseArgs():
    print("ExtractBitlockerKeys.py v%s - by @podalirius_\n" % VERSION)

    parser = argparse.ArgumentParser(description="")

    parser.add_argument("-v", "--verbose", default=False, action="store_true", help='Verbose mode. (default: False)')
    parser.add_argument("-q", "--quiet", dest="quiet", action="store_true", default=False, help="Show no information at all.")
    parser.add_argument("-t", "--threads", dest="threads", action="store", type=int, default=4, required=False, help="Number of threads (default: 4).")

    output = parser.add_argument_group('Output files')
    output.add_argument("--export-xlsx", dest="export_xlsx", type=str, default=None, required=False, help="Output XLSX file to store the results in.")
    output.add_argument("--export-json", dest="export_json", type=str, default=None, required=False, help="Output JSON file to store the results in.")
    output.add_argument("--export-sqlite", dest="export_sqlite", type=str, default=None, required=False, help="Output SQLITE3 file to store the results in.")

    authconn = parser.add_argument_group('Authentication & connection')
    authconn.add_argument('--dc-ip', required=True, action='store', metavar="ip address", help='IP Address of the domain controller or KDC (Key Distribution Center) for Kerberos. If omitted it will use the domain part (FQDN) specified in the identity parameter')
    authconn.add_argument('--kdcHost', dest="kdcHost", action='store', metavar="FQDN KDC", help='FQDN of KDC for Kerberos.')
    authconn.add_argument("-d", "--domain", dest="auth_domain", metavar="DOMAIN", action="store", default="", help="(FQDN) domain to authenticate to")
    authconn.add_argument("-u", "--user", dest="auth_username", metavar="USER", action="store", default="", help="user to authenticate with")

    secret = parser.add_argument_group("Credentials")
    cred = secret.add_mutually_exclusive_group()
    cred.add_argument("--no-pass", default=False, action="store_true", help="Don't ask for password (useful for -k)")
    cred.add_argument("-p", "--password", dest="auth_password", metavar="PASSWORD", action="store", default=None, help="Password to authenticate with")
    cred.add_argument("-H", "--hashes", dest="auth_hashes", action="store", metavar="[LMHASH:]NTHASH", help='NT/LM hashes, format is LMhash:NThash')
    cred.add_argument("--aes-key", dest="auth_key", action="store", metavar="hex key", help='AES key to use for Kerberos Authentication (128 or 256 bits)')
    secret.add_argument("-k", "--kerberos", dest="use_kerberos", action="store_true", help='Use Kerberos authentication. Grabs credentials from .ccache file (KRB5CCNAME) based on target parameters. If valid credentials cannot be found, it will use the ones specified in the command line')

    return parser.parse_args()


if __name__ == '__main__':
    options = parseArgs()
    if options.auth_hashes is not None:
        if ":" not in options.auth_hashes:
            options.auth_hashes = ":" + options.auth_hashes
    auth_lm_hash, auth_nt_hash = parse_lm_nt_hashes(options.auth_hashes)

    if options.auth_key is not None:
        options.use_kerberos = True
    
    if options.use_kerberos is True and options.kdcHost is None:
        print("[!] Specify KDC's Hostname of FQDN using the argument --kdcHost")
        exit()
    
    if not options.quiet:
        print("[>] Extracting BitLocker recovery keys of all computers ...")

    computer_keys = raw_ldap_query(
        auth_domain=options.auth_domain,
        auth_dc_ip=options.dc_ip,
        auth_username=options.auth_username,
        auth_password=options.auth_password,
        auth_hashes=options.auth_hashes,
        auth_key=options.auth_key,
        query="(objectClass=msFVE-RecoveryInformation)",
        attributes=[
            "msFVE-KeyPackage",  # https://learn.microsoft.com/en-us/windows/win32/adschema/a-msfve-keypackage
            "msFVE-RecoveryGuid",  # https://learn.microsoft.com/en-us/windows/win32/adschema/a-msfve-recoveryguid
            "msFVE-RecoveryPassword",  # https://learn.microsoft.com/en-us/windows/win32/adschema/a-msfve-recoverypassword
            "msFVE-VolumeGuid"  # https://learn.microsoft.com/en-us/windows/win32/adschema/a-msfve-volumeguid
        ],
        use_kerberos=options.use_kerberos,
        kdcHost=options.kdcHost
    )

    if not options.quiet:
        print("[>] Found %d BitLocker recovery keys!" % len(computer_keys.keys()))

    results = {}

    if len(computer_keys.keys()) != 0:
        for dn, fve_entry in computer_keys.items():
            if len(fve_entry.keys()) != 0:
                if dn not in results.keys():
                    results[dn] = []
                result = parse_fve(dn, fve_entry)
                print("| %-20s | %-20s | %s |" % (result["domain"], result["computerName"], result["recoveryKey"]))
                results[dn].append(result)

        print("[>] Extracted %d BitLocker recovery keys!" % len(computer_keys.keys()))

        # Export results
        if options.export_json is not None:
            export_json(options, results)

        if options.export_xlsx is not None:
            export_xlsx(options, results)

        if options.export_sqlite is not None:
            export_sqlite(options, results)
    else:
        print("[!] No computers in the domain found matching filter (objectClass=msFVE-RecoveryInformation)")


    print("[+] Bye Bye!")

File: go/src/ExtractBitlockerKeys.go
package main

import (
	"errors"
	"flag"
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/go-ldap/ldap/v3"
	"github.com/xuri/excelize/v2"
)

func banner() {
	fmt.Printf("ExtractBitlockerKeys v%s - by Remi GASCOU (Podalirius)\n", "1.3")
	fmt.Println("")
}

func ldap_init_connection(host string, port int, username string, domain string, password string) (*ldap.Conn, error) {
	// Check if TCP port is valid
	if port < 1 || port > 65535 {
		fmt.Println("[!] Invalid port number. Port must be in the range 1-65535.")
		return nil, errors.New("invalid port number")
	}

	// Set up LDAP connection
	ldapSession, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", host, port))
	if err != nil {
		fmt.Println("[!] Error connecting to LDAP server:", err)
		return nil, nil
	}

	// Bind with credentials if provided
	bindDN := ""
	if username != "" {
		bindDN = fmt.Sprintf("%s@%s", username, domain)
	}
	if bindDN != "" && password != "" {
		err = ldapSession.Bind(bindDN, password)
		if err != nil {
			fmt.Println("[!] Error binding:", err)
			return nil, nil
		}
	}

	return ldapSession, nil
}

func ldap_get_rootdse(ldapSession *ldap.Conn) *ldap.Entry {
	// Specify LDAP search parameters
	// https://pkg.go.dev/gopkg.in/ldap.v3#NewSearchRequest
	searchRequest := ldap.NewSearchRequest(
		// Base DN blank
		"",
		// Scope Base
		ldap.ScopeBaseObject,
		// DerefAliases
		ldap.NeverDerefAliases,
		// SizeLimit
		1,
		// TimeLimit
		0,
		// TypesOnly
		false,
		// Search filter
		"(objectClass=*)",
		// Attributes to retrieve
		[]string{"*"},
		// Controls
		nil,
	)

	// Perform LDAP search
	searchResult, err := ldapSession.Search(searchRequest)
	if err != nil {
		fmt.Println("[!] Error searching LDAP:", err)
		return nil
	}

	return searchResult.Entries[0]
}

func getDomainFromDistinguishedName(distinguishedName string) string {
	domain := ""
	if strings.Contains(strings.ToLower(distinguishedName), "dc=") {
		dnParts := strings.Split(strings.ToLower(distinguishedName), ",")
		for _, part := range dnParts {
			if strings.HasPrefix(part, "dc=") {
				dcValue := strings.SplitN(part, "=", 2)[1]
				if domain == "" {
					domain = dcValue
				} else {
					domain = domain + "." + dcValue
				}
			}
		}
	}
	return domain
}

func getOUPathFromDistinguishedName(distinguishedName string) string {
	ouPath := ""
	if strings.Contains(strings.ToLower(distinguishedName), "ou=") {
		dnParts := strings.Split(strings.ToLower(distinguishedName), ",")
		// Reverse dnParts slice
		for i, j := 0, len(dnParts)-1; i < j; i, j = i+1, j-1 {
			dnParts[i], dnParts[j] = dnParts[j], dnParts[i]
		}

		// Skip domain
		for len(dnParts) > 0 && strings.HasPrefix(dnParts[0], "dc=") {
			dnParts = dnParts[1:]
		}

		for len(dnParts) > 0 && strings.HasPrefix(dnParts[0], "ou=") {
			ouValue := strings.SplitN(dnParts[0], "=", 2)[1]
			if ouPath == "" {
				ouPath = ouValue
			} else {
				ouPath = ouPath + " --> " + ouValue
			}
			dnParts = dnParts[1:]
		}
	}
	return ouPath
}

func parseFVE(distinguishedName string, ldapEntry *ldap.Entry) map[string]string {
	entry := make(map[string]string)
	entry["distinguishedName"] = distinguishedName
	entry["domain"] = getDomainFromDistinguishedName(distinguishedName)
	entry["organizationalUnits"] = getOUPathFromDistinguishedName(distinguishedName)
	entry["createdAt"] = ""
	entry["volumeGuid"] = ""

	// Parse CN of key
	re := regexp.MustCompile(`^(CN=)([0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}-[0-9]{2}:[0-9]{2})({[0-9A-F\-]+}),`)
	matched := re.FindStringSubmatch(distinguishedName)
	if matched != nil {
		createdAt, guid := matched[2], matched[3]
		entry["createdAt"] = createdAt
		entry["volumeGuid"] = strings.Trim(guid, "{}")
	}

	// Parse computer name
	entry["computerName"] = ""
	if strings.Contains(distinguishedName, ",") {
		splitDN := strings.Split(distinguishedName, ",")
		if strings.ToUpper(splitDN[1][:3]) == "CN=" {
			entry["computerName"] = strings.SplitN(splitDN[1], "=", 2)[1]
		}
	}

	// Add recovery key
	entry["recoveryKey"] = ldapEntry.GetAttributeValue("msFVE-RecoveryPassword")

	return entry
}


var (
	useLdaps     bool
	quiet        bool
	debug        bool
	ldapHost     string
	ldapPort     int
	authDomain   string
	authUsername string
	// noPass         bool
	authPassword string
	authHashes   string
	// authKey        string
	// useKerberos    bool
	xlsx         string
)

func parseArgs() {
	flag.BoolVar(&useLdaps, "use-ldaps", false, "Use LDAPS instead of LDAP.")
	flag.BoolVar(&quiet, "quiet", false, "Show no information at all.")
	flag.BoolVar(&debug, "debug", false, "Debug mode")

	flag.StringVar(&ldapHost, "host", "", "IP Address of the domain controller or KDC (Key Distribution Center) for Kerberos. If omitted it will use the domain part (FQDN) specified in the identity parameter.")
	flag.IntVar(&ldapPort, "port", 0, "Port number to connect to LDAP server.")

	flag.StringVar(&authDomain, "domain", "", "(FQDN) domain to authenticate to.")
	flag.StringVar(&authUsername, "username", "", "User to authenticate as.")
	//flag.BoolVar(&noPass, "no-pass", false, "don't ask for password (useful for -k)")
	flag.StringVar(&authPassword, "password", "", "password to authenticate with.")
	flag.StringVar(&authHashes, "hashes", "", "NT/LM hashes, format is LMhash:NThash.")
	//flag.StringVar(&authKey, "aes-key", "", "AES key to use for Kerberos Authentication (128 or 256 bits)")
	//flag.BoolVar(&useKerberos, "k", false, "Use Kerberos authentication. Grabs credentials from .ccache file (KRB5CCNAME) based on target parameters. If valid credentials cannot be found, it will use the ones specified in the command line")

	flag.StringVar(&xlsx, "xlsx", "", "Output results in a XLSX Excel file.")

	flag.Parse()

	if ldapHost == "" {
		fmt.Println("[!] Option -host <host> is required.")
		flag.Usage()
		os.Exit(1)
	}

	if ldapPort == 0 {
		if useLdaps {
			ldapPort = 636
		} else {
			ldapPort = 389
		}
	}
}

func main() {
	banner()
	parseArgs()

	if debug {
		if !useLdaps {
			fmt.Printf("[debug] Connecting to remote ldap://%s:%d ...\n", ldapHost, ldapPort)
		} else {
			fmt.Printf("[debug] Connecting to remote ldaps://%s:%d ...\n", ldapHost, ldapPort)
		}
	}

	// Init the LDAP connection
	ldapSession, err := ldap_init_connection(ldapHost, ldapPort, authUsername, authDomain, authPassword)
	if err != nil {
		fmt.Println("[!] Error searching LDAP:", err)
		return
	}

	rootDSE := ldap_get_rootdse(ldapSession)
	if debug {
		fmt.Printf("[debug] Using defaultNamingContext %s ...\n", rootDSE.GetAttributeValue("defaultNamingContext"))
	}

	// Specify LDAP search parameters
	// https://pkg.go.dev/gopkg.in/ldap.v3#NewSearchRequest
	searchRequest := ldap.NewSearchRequest(
		// Base DN
		rootDSE.GetAttributeValue("defaultNamingContext"),
		// Scope
		ldap.ScopeWholeSubtree,
		// DerefAliases
		ldap.NeverDerefAliases,
		// SizeLimit
		0,
		// TimeLimit
		0,
		// TypesOnly
		false,
		// Search filter
		"(objectClass=msFVE-RecoveryInformation)",
		// Attributes to retrieve
		[]string{
			"msFVE-KeyPackage",        // https://learn.microsoft.com/en-us/windows/win32/adschema/a-msfve-keypackage
			"msFVE-RecoveryGuid",      // https://learn.microsoft.com/en-us/windows/win32/adschema/a-msfve-recoveryguid
			"msFVE-RecoveryPassword",  // https://learn.microsoft.com/en-us/windows/win32/adschema/a-msfve-recoverypassword
			"msFVE-VolumeGuid",        // https://learn.microsoft.com/en-us/windows/win32/adschema/a-msfve-volumeguid
			"distinguishedName",
		},
		// Controls
		nil,
	)

	// Perform LDAP search
	fmt.Println("[+] Extracting LAPS passwords of all computers ... ")
	searchResult, err := ldapSession.Search(searchRequest)
	if err != nil {
		fmt.Println("[!] Error searching LDAP:", err)
		return
	}
	
	// Print search results
	var resultsList []map[string]string
	for _, entry := range searchResult.Entries {
		result := parseFVE(entry.GetAttributeValue("distinguishedName"), entry)
		resultsList = append(resultsList, result)
	}
	fmt.Printf("[+] Total BitLocker recovery keys found: %d\n", len(resultsList))
	
	// Export BitLocker Recovery Keys to an Excel
	if xlsx != "" {
		f := excelize.NewFile()
		// Create a new sheet.
		index, err := f.NewSheet("Sheet1")
		// Set value of a cell.
		f.SetCellValue("Sheet1", "A1", "Domain")
		f.SetCellValue("Sheet1", "B1", "Computer Name")
		f.SetCellValue("Sheet1", "C1", "BitLocker Recovery Key")
		for i, result := range resultsList {
			f.SetCellValue("Sheet1", fmt.Sprintf("A%d", i+2), result["domain"])
			f.SetCellValue("Sheet1", fmt.Sprintf("B%d", i+2), result["computerName"])
			f.SetCellValue("Sheet1", fmt.Sprintf("C%d", i+2), result["recoveryKey"])
		}
		// Set active sheet of the workbook.
		f.SetActiveSheet(index)
		// Save xlsx file by the given path.
		if err := f.SaveAs(xlsx); err != nil {
			fmt.Println(err)
		}
		fmt.Printf("[+] Exported BitLocker recovery keys to: %s\n", xlsx)
	} else {
		// Print the keys in the console
		for _, result := range resultsList {
			fmt.Printf("| %-20s | %-20s | %s |\n", result["domain"], result["computerName"], result["recoveryKey"])
		}
	}

	fmt.Println("[+] All done!")
}
